jit-compiler/src/codegen.rs

```diff
@@ -0,0 +1,363 @@
+use std::collections::HashMap;
+
+use itertools::Itertools;
+use powdr_ast::{
+    analyzed::{Analyzed, Expression, FunctionValueDefinition, PolynomialReference, Reference},
+    parsed::{
+        display::{format_type_args, quote},
+        types::{ArrayType, FunctionType, Type, TypeScheme},
+        ArrayLiteral, BinaryOperation, BinaryOperator, BlockExpression, FunctionCall, IfExpression,
+        IndexAccess, LambdaExpression, Number, StatementInsideBlock, UnaryOperation,
+    },
+};
+use powdr_number::FieldElement;
+
+pub struct CodeGenerator<'a, T> {
+    analyzed: &'a Analyzed<T>,
+    /// Symbols mapping to either their code or an error message explaining
+    /// why they could not be compiled.
+    /// While the code is still being generated, this contains `None`.
+    symbols: HashMap<String, Result<Option<String>, String>>,
+}
+
+impl<'a, T: FieldElement> CodeGenerator<'a, T> {
+    pub fn new(analyzed: &'a Analyzed<T>) -> Self {
+        Self {
+            analyzed,
+            symbols: Default::default(),
+        }
+    }
+
+    /// Request a symbol to be compiled. The code can later be retrieved
+    /// via `compiled_symbols`.
+    /// In the error case, `self` can still be used to compile other symbols.
+    pub fn request_symbol(&mut self, name: &str) -> Result<(), String> {
+        match self.symbols.get(name) {
+            Some(Ok(_)) => Ok(()),
+            Some(Err(e)) => Err(e.clone()),
+            None => {
+                let name = name.to_string();
+                self.symbols.insert(name.clone(), Ok(None));
+                let to_insert;
+                let to_return;
+                match self.generate_code(&name) {
+                    Ok(code) => {
+                        to_insert = Ok(Some(code));
+                        to_return = Ok(());
+                    }
+                    Err(err) => {
+                        to_insert = Err(err.clone());
+                        to_return = Err(err);
+                    }
+                }
```

Comment:
```suggestion
                let (to_insert, to_return) =
                match self.generate_code(&name) {
                    Ok(code) => {
                        (Ok(Some(code)), Ok(()))
                    }
                    Err(err) => {
                        (Err(err.clone()), Err(err))
                    }
                };
```

Reply:
The next PR heavily modifies this again anyway, so I would just keep it for now: https://github.com/powdr-labs/powdr/pull/1837/files#diff-adff81349be67708164c673cd6da23098e721c006cd10cf61add179bb33b0679R37

Reply:
Fine